File: python/bench_gpu.py
# SPDX-License-Identifier: CC-BY-NC-SA-4.0
# Copyright (c) 2025-2026 fumi-engineer

"""GPU-only (Metal) benchmark for Python MoE Transformer.

Imports from python.gpu -- requires PyObjC Metal/MPS on macOS.

Measures:
  1. GPU kernel micro-benchmarks (MPS matmul, MSL softmax, MSL rmsnorm)
  2. Full-model GPU forward at 3 scales
  3. Full-model GPU train proxy (forward + CE loss sum, no readback) at 3 scales

Policy:
  - Setup uploads model/input/target buffers once.
  - Timed loops do not permit CPU<->GPU transfer APIs.
"""

import gc
import json
import os
import platform
import resource
import sys
import time
import tracemalloc
from datetime import datetime, timezone
from statistics import median

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np


def _log(msg):
    print(msg, file=sys.stderr)


# Check Metal availability before proceeding
try:
    from python.gpu import (
        metal_available,
        MetalContext,
        MetalTensor,
        MetalMoETransformer,
        MetalTrainer,
        seed_rng as gpu_seed_rng,
    )
    from python.config import Config

    if not metal_available():
        _log("Metal not available on this system. Exiting.")
        sys.exit(0)
except ImportError as e:
    _log(f"GPU module import failed: {e}")
    _log("Install: pip install pyobjc-framework-Metal pyobjc-framework-MetalPerformanceShaders")
    sys.exit(1)


def _env_int(name: str, default: int) -> int:
    raw = os.environ.get(name)
    if raw is None:
        return default
    try:
        val = int(raw)
    except ValueError:
        return default
    return val if val > 0 else default


_ORIG_TO_NUMPY = MetalTensor.to_numpy
_ORIG_FROM_NUMPY = MetalTensor.__dict__["from_numpy"]
_ORIG_NUMPY_VIEW = MetalTensor.numpy_view


def _install_transfer_guard() -> None:
    """Ban host transfer APIs during warmup/timed loops."""
    def _blocked_to_numpy(*_args, **_kwargs):
        raise RuntimeError("CPU<->GPU transfer is forbidden in bench_gpu.py timed loop (to_numpy)")

    def _blocked_from_numpy(*_args, **_kwargs):
        raise RuntimeError("CPU<->GPU transfer is forbidden in bench_gpu.py timed loop (from_numpy)")

    def _blocked_numpy_view(*_args, **_kwargs):
        raise RuntimeError("CPU<->GPU transfer is forbidden in bench_gpu.py timed loop (numpy_view)")

    MetalTensor.to_numpy = _blocked_to_numpy
    MetalTensor.from_numpy = classmethod(_blocked_from_numpy)
    MetalTensor.numpy_view = _blocked_numpy_view


def _remove_transfer_guard() -> None:
    MetalTensor.to_numpy = _ORIG_TO_NUMPY
    MetalTensor.from_numpy = _ORIG_FROM_NUMPY
    MetalTensor.numpy_view = _ORIG_NUMPY_VIEW


# Constants
N_TRIALS = _env_int("ROSETTA_BENCH_TRIALS", 10)
N_WARMUP = _env_int("ROSETTA_BENCH_WARMUP", 3)
SEED = 42
VOCAB = 1000

# GC tracking
_gc_time_acc = 0
_gc_start_time = 0


def _gc_callback(phase, info):
    global _gc_time_acc, _gc_start_time
    if phase == "start":
        _gc_start_time = time.perf_counter_ns()
    elif phase == "stop":
        _gc_time_acc += time.perf_counter_ns() - _gc_start_time


gc.callbacks.append(_gc_callback)


def _percentile(sorted_vals, p):
    n = len(sorted_vals)
    k = (n - 1) * p / 100.0
    f = int(k)
    c = f + 1
    if c >= n:
        return sorted_vals[-1]
    return sorted_vals[f] + (k - f) * (sorted_vals[c] - sorted_vals[f])


def _measure(setup_fn, run_fn, scenario_id, n_warmup=N_WARMUP, n_trials=N_TRIALS):
    global _gc_time_acc
    _log(f"  [{scenario_id}] setup...")
    ctx = setup_fn()
    _install_transfer_guard()
    try:
        _log(f"  [{scenario_id}] warmup x{n_warmup}")
        warmup_timings_ns = []
        for _ in range(n_warmup):
            t0 = time.perf_counter_ns()
            run_fn(ctx)
            t1 = time.perf_counter_ns()
            warmup_timings_ns.append(t1 - t0)

        gc.collect()
        gc_pause_before = sum(s.get("collections", 0) for s in gc.get_stats())
        _gc_time_acc = 0
        tracemalloc.start()

        timings_ns = []
        cpu_times_ns = []
        last_output = None
        _log(f"  [{scenario_id}] measuring x{n_trials}")
        for _ in range(n_trials):
            ru_before = resource.getrusage(resource.RUSAGE_SELF)
            t0 = time.perf_counter_ns()
            last_output = run_fn(ctx)
            t1 = time.perf_counter_ns()
            ru_after = resource.getrusage(resource.RUSAGE_SELF)
            timings_ns.append(t1 - t0)
            cpu_ns = int(
                ((ru_after.ru_utime + ru_after.ru_stime)
                 - (ru_before.ru_utime + ru_before.ru_stime)) * 1e9
            )
            cpu_times_ns.append(cpu_ns)
    finally:
        _remove_transfer_guard()

    gc_wall_time_ns = _gc_time_acc
    rss = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    gc_pause_after = sum(s.get("collections", 0) for s in gc.get_stats())
    tracemalloc.stop()

    sorted_t = sorted(timings_ns)
    gc_collections = gc_pause_after - gc_pause_before
    q1 = int(_percentile(sorted_t, 25))
    q3 = int(_percentile(sorted_t, 75))
    median_ns = int(median(timings_ns))
    cpu_median_ns = int(median(cpu_times_ns))
    median_sec = median_ns / 1e9
    sum_timings = sum(timings_ns)
    gc_throughput = 1.0 - (gc_wall_time_ns / sum_timings) if sum_timings > 0 else 1.0

    # Check numerical output
    nan_count = inf_count = 0
    max_abs = 0.0
    if isinstance(last_output, np.ndarray):
        nan_count = int(np.isnan(last_output).sum())
        inf_count = int(np.isinf(last_output).sum())
        finite = last_output[np.isfinite(last_output)]
        max_abs = float(np.max(np.abs(finite))) if finite.size > 0 else 0.0

    return {
        "timings_ns": timings_ns,
        "cpu_times_ns": cpu_times_ns,
        "warmup_timings_ns": warmup_timings_ns,
        "median_ns": median_ns,
        "p95_ns": int(_percentile(sorted_t, 95)),
        "min_ns": min(timings_ns),
        "max_ns": max(timings_ns),
        "iqr_ns": q3 - q1,
        "cpu_median_ns": cpu_median_ns,
        "memory": {"peak_rss_bytes": rss},
        "gc": {"total_gc_time_ns": gc_wall_time_ns, "gc_pause_count": gc_collections},
        "numerical": {"nan_count": nan_count, "inf_count": inf_count, "max_abs": max_abs},
        "derived": {"gc_throughput": gc_throughput},
    }


def _add_throughput(result, batch, seq_len):
    median_sec = result["median_ns"] / 1e9
    if median_sec > 0:
        result["throughput_tokens_per_sec"] = (batch * seq_len) / median_sec


# --- GPU Kernel Benchmarks ---

def scenario_gpu_kernel_matmul():
    """GPU MPS matmul 256x256."""
    m, n, k = 256, 256, 256
    known_flops = 2 * m * n * k

    def setup():
        ctx = MetalContext()
        a = MetalTensor.empty(ctx, [m, k])
        b = MetalTensor.empty(ctx, [k, n])
        c = MetalTensor.empty(ctx, [m, n])
        return ctx, a, b, c

    def run(ctx_tuple):
        ctx, a, b, c = ctx_tuple
        ctx.mps_matmul(a, b, c, m, n, k)
        return None

    result = _measure(setup, run, "gpu_kernel_matmul")
    median_sec = result["median_ns"] / 1e9
    gflops = (known_flops / median_sec / 1e9) if median_sec > 0 else 0.0
    result["derived"]["gflops"] = gflops
    result.update(id="gpu_kernel_matmul", axis="gpu",
                  params={"m": m, "n": n, "k": k}, known_flops=known_flops,
                  warmup_runs=N_WARMUP, trial_runs=N_TRIALS)
    return result


def scenario_gpu_kernel_softmax():
    """GPU MSL softmax kernel."""
    n = 1000

    def setup():
        ctx = MetalContext()
        inp = MetalTensor.empty(ctx, [1, n])
        out = MetalTensor.empty(ctx, [1, n])
        return ctx, inp, out

    def run(ctx_tuple):
        ctx, inp, out = ctx_tuple
        ctx.dispatch_kernel("softmax", [inp, out, n], grid_size=1, threadgroup_size=256)
        return None

    result = _measure(setup, run, "gpu_kernel_softmax")
    result.update(id="gpu_kernel_softmax", axis="gpu",
                  params={"n": n}, warmup_runs=N_WARMUP, trial_runs=N_TRIALS)
    return result


def scenario_gpu_kernel_rmsnorm():
    """GPU MSL rmsnorm kernel."""
    rows, hidden = 2, 64

    def setup():
        ctx = MetalContext()
        inp = MetalTensor.empty(ctx, [rows, hidden])
        w = MetalTensor.empty(ctx, [hidden])
        out = MetalTensor.empty(ctx, [rows, hidden])
        return ctx, inp, w, out

    def run(ctx_tuple):
        ctx, inp, w, out = ctx_tuple
        ctx.dispatch_kernel("rmsnorm", [inp, out, w, hidden, 1e-6],
                            grid_size=rows * 256, threadgroup_size=256)
        return None

    result = _measure(setup, run, "gpu_kernel_rmsnorm")
    result.update(id="gpu_kernel_rmsnorm", axis="gpu",
                  params={"rows": rows, "hidden_dim": hidden},
                  warmup_runs=N_WARMUP, trial_runs=N_TRIALS)
    return result


# --- GPU Forward/Train ---

def _gpu_forward_scenario(label, hidden_dim, cfg_method):
    """Full-model GPU forward at given scale."""
    batch, seq = 2, 32

    def setup():
        gpu_seed_rng(SEED)
        cfg = cfg_method()
        ctx = MetalContext()
        model = MetalMoETransformer.from_config(ctx, cfg)
        token_ids = (np.arange(batch * seq, dtype=np.float32) % VOCAB).reshape(batch, seq)
        token_ids_mt = MetalTensor(ctx, data=token_ids)
        return model, token_ids_mt, batch, seq

    def run(ctx_tuple):
        model, token_ids_mt, b, s = ctx_tuple
        model.forward_tokens_tensor(token_ids_mt, b, s)
        return None

    result = _measure(setup, run, f"gpu_forward_{label}")
    result.update(id=f"gpu_forward_{label}", axis="gpu",
                  params={"batch": batch, "seq_len": seq, "hidden_dim": hidden_dim},
                  warmup_runs=N_WARMUP, trial_runs=N_TRIALS)
    _add_throughput(result, batch, seq)
    return result


def _gpu_train_scenario(label, hidden_dim, cfg_method):
    """Full-model GPU train proxy at given scale (forward + CE loss sum)."""
    batch, seq = 2, 8

    def setup():
        gpu_seed_rng(SEED)
        cfg = cfg_method()
        ctx = MetalContext()
        model = MetalMoETransformer.from_config(ctx, cfg)
        trainer = MetalTrainer(model)
        token_ids = (np.arange(batch * seq, dtype=np.float32) % VOCAB).reshape(batch, seq)
        targets = ((np.arange(batch * seq, dtype=np.float32) + 1.0) % VOCAB)
        token_ids_mt = MetalTensor(ctx, data=token_ids)
        targets_mt = MetalTensor(ctx, data=targets)
        return trainer, token_ids_mt, targets_mt, batch, seq

    def run(ctx_tuple):
        trainer, token_ids_mt, targets_mt, b, s = ctx_tuple
        trainer.train_step_gpu_tensors(
            token_ids_mt,
            targets_mt,
            batch=b,
            seq_len=s,
            readback=False,
        )
        return None

    result = _measure(setup, run, f"gpu_train_{label}")
    result.update(id=f"gpu_train_{label}", axis="gpu",
                  params={"batch": batch, "seq_len": seq, "hidden_dim": hidden_dim},
                  warmup_runs=N_WARMUP, trial_runs=N_TRIALS)
    _add_throughput(result, batch, seq)
    return result


def _metadata():
    return {
        "language": "python",
        "backend": "gpu_metal",
        "language_version": platform.python_version(),
        "os": platform.system() + " " + platform.release(),
        "cpu_model": platform.processor(),
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "seed": SEED,
        "n_trials": N_TRIALS,
        "n_warmup": N_WARMUP,
    }


def main():
    _log("Python MoE Benchmark -- GPU-only Metal (no host transfer in timed loop)")
    np.random.seed(SEED)

    scenarios = []
    total = 9
    idx = 0

    # Kernel benchmarks
    idx += 1
    _log(f"[{idx}/{total}] gpu_kernel_matmul")
    scenarios.append(scenario_gpu_kernel_matmul())

    idx += 1
    _log(f"[{idx}/{total}] gpu_kernel_softmax")
    scenarios.append(scenario_gpu_kernel_softmax())

    idx += 1
    _log(f"[{idx}/{total}] gpu_kernel_rmsnorm")
    scenarios.append(scenario_gpu_kernel_rmsnorm())

    # Forward pass
    idx += 1
    _log(f"[{idx}/{total}] gpu_forward_64")
    scenarios.append(_gpu_forward_scenario("64", 64, Config.tiny))

    idx += 1
    _log(f"[{idx}/{total}] gpu_forward_256")
    scenarios.append(_gpu_forward_scenario("256", 256, Config.small))

    idx += 1
    _log(f"[{idx}/{total}] gpu_forward_512")
    scenarios.append(_gpu_forward_scenario("512", 512, Config.medium))

    # Train step
    idx += 1
    _log(f"[{idx}/{total}] gpu_train_64")
    scenarios.append(_gpu_train_scenario("64", 64, Config.tiny))

    idx += 1
    _log(f"[{idx}/{total}] gpu_train_256")
    scenarios.append(_gpu_train_scenario("256", 256, Config.small))

    idx += 1
    _log(f"[{idx}/{total}] gpu_train_512")
    scenarios.append(_gpu_train_scenario("512", 512, Config.medium))

    result = {"metadata": _metadata(), "scenarios": scenarios}
    _log("Done. Writing JSON to stdout.")
    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: julia/bench_gpu.jl
# SPDX-License-Identifier: CC-BY-NC-SA-4.0
# Copyright (c) 2025-2026 fumi-engineer

# bench_gpu.jl — Metal GPU benchmark harness for MoE Transformer
#
# Runs GPU-only scenarios using MetalTensor (MtlArray).
# Includes from src_gpu/ only (Metal.jl required).
#
# Usage: julia --project=. bench_gpu.jl

include("src_gpu/MoETransformerGPU.jl")
using .MoETransformerGPU
using Metal
using LinearAlgebra: mul!
using Random
using Dates
using Printf

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------
function _env_int(name::String, default::Int)::Int
    raw = get(ENV, name, string(default))
    parsed = try
        parse(Int, raw)
    catch
        default
    end
    parsed > 0 ? parsed : default
end

const N_TRIALS  = _env_int("ROSETTA_BENCH_TRIALS", 10)
const N_WARMUP  = _env_int("ROSETTA_BENCH_WARMUP", 3)
const SEED      = 42
const VOCAB     = 1000
const HIDDEN    = 64

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------
function get_cpu_model()
    if Sys.isapple()
        try return strip(read(`sysctl -n machdep.cpu.brand_string`, String)) catch; end
    end
    "unknown"
end

function make_forward_input_mtl(batch::Int, seq_len::Int)
    data = Array{Float32}(undef, batch, seq_len)
    for b in 1:batch, s in 1:seq_len
        data[b, s] = Float32(((b - 1) * seq_len + (s - 1)) % VOCAB)
    end
    MetalTensor(MtlArray(data), F32)
end

function make_targets_mtl(batch::Int, seq_len::Int)
    data = Array{Float32}(undef, batch, seq_len)
    for b in 1:batch, s in 1:seq_len
        data[b, s] = Float32(((b - 1) * seq_len + (s - 1) + 1) % VOCAB)
    end
    MetalTensor(MtlArray(data), F32)
end

function check_numerical_mtl(t::MetalTensor)
    d = t.data
    nan_count = Int(round(sum(Float32.(isnan.(d)))))
    inf_count = Int(round(sum(Float32.(isinf.(d)))))
    max_abs = Float32(maximum(abs.(ifelse.(isfinite.(d), d, 0f0))))
    (nan_count, inf_count, max_abs)
end

function median_val(v::Vector{UInt64})
    s = sort(v)
    n = length(s)
    n == 0 && return UInt64(0)
    if n % 2 == 1
        s[div(n, 2) + 1]
    else
        div(s[div(n, 2)] + s[div(n, 2) + 1], UInt64(2))
    end
end

function p95_val(v::Vector{UInt64})
    s = sort(v)
    idx = Int(ceil(0.95 * length(s)))
    s[min(idx, length(s))]
end

function percentile_val(v::Vector{UInt64}, p::Float64)
    s = sort(v)
    n = length(s)
    k = (n - 1) * p / 100.0
    f = Int(floor(k)) + 1
    c = f + 1
    c > n && return s[n]
    UInt64(round(Float64(s[f]) + (k - (f - 1)) * Float64(s[c] - s[f])))
end

function iqr_val(v::Vector{UInt64})
    q1 = percentile_val(v, 25.0)
    q3 = percentile_val(v, 75.0)
    q3 - q1
end

function escape_json(s::AbstractString)
    buf = IOBuffer()
    for c in s
        if c == '"'; write(buf, "\\\"")
        elseif c == '\\'; write(buf, "\\\\")
        elseif c == '\n'; write(buf, "\\n")
        elseif c == '\r'; write(buf, "\\r")
        elseif c == '\t'; write(buf, "\\t")
        else write(buf, c)
        end
    end
    String(take!(buf))
end

format_timings(timings::Vector{UInt64}) = join(string.(timings), ", ")

function get_cpu_time_ns()::UInt64
    buf = zeros(UInt8, 256)
    ret = ccall(:getrusage, Cint, (Cint, Ptr{UInt8}), 0, buf)
    ret != 0 && return UInt64(0)
    utime_sec  = reinterpret(Int64, buf[1:8])[1]
    utime_usec = reinterpret(Int64, buf[9:16])[1]
    stime_sec  = reinterpret(Int64, buf[17:24])[1]
    stime_usec = reinterpret(Int64, buf[25:32])[1]
    total_usec = UInt64(utime_sec * 1_000_000 + utime_usec + stime_sec * 1_000_000 + stime_usec)
    total_usec * UInt64(1000)
end

mutable struct ScenarioResult
    id::String
    axis::String
    params::Dict{String,Any}
    n_warmup::Int
    n_trials::Int
    timings_ns::Vector{UInt64}
    cpu_times_ns::Vector{UInt64}
    warmup_timings_ns::Vector{UInt64}
    median_ns::UInt64
    p95_ns::UInt64
    min_ns::UInt64
    max_ns::UInt64
    iqr_ns::UInt64
    cpu_median_ns::UInt64
    throughput_tokens_per_sec::Float64
    peak_rss_bytes::Int64
    alloc_bytes::Int64
    gc_total_time_ns::Int64
    gc_pause_count::Int64
    nan_count::Int
    inf_count::Int
    max_abs::Float32
    alloc_rate_bytes_per_sec::Float64
    gc_throughput::Float64
    gflops::Union{Float64,Nothing}
end

function run_scenario(;
    id::String,
    axis::String,
    params::Dict{String,Any},
    setup_fn::Function,
    run_fn::Function,
    n_warmup::Int = N_WARMUP,
    n_trials::Int = N_TRIALS,
    precompile::Bool = false,
    known_flops::Union{Int,Nothing} = nothing
)::ScenarioResult
    print(stderr, "  $(id)...")
    flush(stderr)

    if precompile
        pre_ctx = setup_fn()
        Metal.@sync run_fn(pre_ctx)
    end

    ctx = setup_fn()

    warmup_timings = Vector{UInt64}(undef, n_warmup)
    for i in 1:n_warmup
        tw0 = time_ns()
        Metal.@sync run_fn(ctx)
        tw1 = time_ns()
        warmup_timings[i] = UInt64(tw1 - tw0)
    end

    GC.gc()
    gc_before = Base.gc_num()

    timings = Vector{UInt64}(undef, n_trials)
    cpu_times = Vector{UInt64}(undef, n_trials)
    last_result = nothing

    for i in 1:n_trials
        cpu0 = get_cpu_time_ns()
        t0 = time_ns()
        result = Metal.@sync run_fn(ctx)
        t1 = time_ns()
        cpu1 = get_cpu_time_ns()
        timings[i] = UInt64(t1 - t0)
        cpu_times[i] = cpu1 - cpu0
        if i == n_trials
            last_result = result
        end
    end

    gc_after = Base.gc_num()

    alloc_bytes = max(Int64(0), gc_after.allocd - gc_before.allocd)
    gc_time = gc_after.total_time - gc_before.total_time
    gc_pauses = gc_after.pause - gc_before.pause
    peak_rss = Sys.maxrss()

    med = n_trials > 0 ? median_val(timings) : UInt64(0)
    p95 = n_trials > 0 ? p95_val(timings) : UInt64(0)
    mn = n_trials > 0 ? minimum(timings) : UInt64(0)
    mx = n_trials > 0 ? maximum(timings) : UInt64(0)
    iqr = n_trials > 0 ? iqr_val(timings) : UInt64(0)
    cpu_med = n_trials > 0 ? median_val(cpu_times) : UInt64(0)

    nan_count = 0
    inf_count = 0
    max_abs = 0f0
    if last_result isa MetalTensor
        nan_count, inf_count, max_abs = check_numerical_mtl(last_result)
    end

    b = get(params, "batch", 0)
    sl = get(params, "seq_len", 0)
    throughput = (b > 0 && sl > 0 && med > 0) ? Float64(b * sl) / (Float64(med) / 1e9) : 0.0

    alloc_rate = med > 0 ? Float64(alloc_bytes) / (Float64(med) * 1e-9) : 0.0
    sum_timings = Float64(sum(timings))
    gc_tp = sum_timings > 0 ? 1.0 - (Float64(gc_time) / sum_timings) : 1.0

    gflops = nothing
    if known_flops !== nothing && med > 0
        gflops = Float64(known_flops) / (Float64(med) * 1e-9) * 1e-9
    end

    println(stderr, " done (median=$(med)ns)")

    ScenarioResult(
        id, axis, params, n_warmup, n_trials,
        timings, cpu_times, warmup_timings,
        med, p95, mn, mx, iqr, cpu_med,
        throughput,
        peak_rss, alloc_bytes,
        gc_time, gc_pauses,
        nan_count, inf_count, max_abs,
        alloc_rate, gc_tp, gflops
    )
end

function scenario_to_json(s::ScenarioResult)
    param_entries = String[]
    for (k, v) in s.params
        if v isa String
            push!(param_entries, "\"$(escape_json(k))\": \"$(escape_json(v))\"")
        elseif v isa AbstractVector
            arr_str = join(string.(v), ", ")
            push!(param_entries, "\"$(escape_json(k))\": [$(arr_str)]")
        else
            push!(param_entries, "\"$(escape_json(k))\": $(v)")
        end
    end
    params_json = join(param_entries, ", ")
    timings_json = format_timings(s.timings_ns)
    cpu_json = format_timings(s.cpu_times_ns)
    warmup_json = format_timings(s.warmup_timings_ns)

    derived_entries = String[]
    push!(derived_entries, "\"alloc_rate_bytes_per_sec\": $(s.alloc_rate_bytes_per_sec)")
    push!(derived_entries, "\"gc_throughput\": $(s.gc_throughput)")
    if s.gflops !== nothing
        push!(derived_entries, "\"gflops\": $(s.gflops)")
    end
    derived_json = join(derived_entries, ", ")

    """{
      "id": "$(escape_json(s.id))",
      "axis": "$(escape_json(s.axis))",
      "params": {$(params_json)},
      "warmup_runs": $(s.n_warmup),
      "trial_runs": $(s.n_trials),
      "timings_ns": [$(timings_json)],
      "cpu_times_ns": [$(cpu_json)],
      "warmup_timings_ns": [$(warmup_json)],
      "median_ns": $(s.median_ns),
      "p95_ns": $(s.p95_ns),
      "min_ns": $(s.min_ns),
      "max_ns": $(s.max_ns),
      "iqr_ns": $(s.iqr_ns),
      "cpu_median_ns": $(s.cpu_median_ns),
      "throughput_tokens_per_sec": $(s.throughput_tokens_per_sec),
      "memory": {"peak_rss_bytes": $(s.peak_rss_bytes), "alloc_bytes": $(s.alloc_bytes)},
      "gc": {"total_gc_time_ns": $(s.gc_total_time_ns), "gc_pause_count": $(s.gc_pause_count)},
      "numerical": {"nan_count": $(s.nan_count), "inf_count": $(s.inf_count), "max_abs": $(s.max_abs)},
      "derived": {$(derived_json)}
    }"""
end

function main()
    if !metal_available()
        println(stderr, "ERROR: Metal GPU not available on this system")
        exit(1)
    end

    println(stderr, "Julia MoE Transformer GPU Benchmark (Metal)")
    println(stderr, "============================================================")
    set_inference_mode!(false)

    scenarios = ScenarioResult[]

    # GPU kernel benchmarks
    println(stderr, "GPU Kernels")

    # gpu_kernel_matmul
    push!(scenarios, run_scenario(
        id="gpu_kernel_matmul", axis="gpu",
        params=Dict{String,Any}("m" => 256, "n" => 256, "k" => 256),
        setup_fn=() -> begin
            a = MtlArray(randn(Float32, 256, 256))
            b = MtlArray(randn(Float32, 256, 256))
            c = MtlArray{Float32}(undef, 256, 256)
            (a=a, b=b, c=c)
        end,
        run_fn=ctx -> begin
            mul!(ctx.c, ctx.a, ctx.b)
            nothing
        end,
        precompile=true,
        known_flops=2*256*256*256
    ))

    # gpu_kernel_softmax
    push!(scenarios, run_scenario(
        id="gpu_kernel_softmax", axis="gpu",
        params=Dict{String,Any}("n" => 1000),
        setup_fn=() -> begin
            x = MtlArray(reshape(randn(Float32, 1000), 1, 1000))
            (x=x,)
        end,
        run_fn=ctx -> begin
            mx = maximum(ctx.x; dims=2)
            shifted = ctx.x .- mx
            e = exp.(shifted)
            s = sum(e; dims=2)
            _ = e ./ s
            nothing
        end,
        precompile=true,
        known_flops=4*1000
    ))

    # gpu_kernel_rmsnorm
    push!(scenarios, run_scenario(
        id="gpu_kernel_rmsnorm", axis="gpu",
        params=Dict{String,Any}("rows" => 2, "hidden_dim" => 64),
        setup_fn=() -> begin
            x = MtlArray(randn(Float32, 2, 64))
            w = MtlArray(ones(Float32, 64))
            (x=x, w=w)
        end,
        run_fn=ctx -> begin
            sum_sq = sum(ctx.x .* ctx.x; dims=2)
            inv_rms = 1f0 ./ sqrt.(sum_sq ./ 64f0 .+ 1f-6)
            _ = ctx.x .* inv_rms .* reshape(ctx.w, 1, 64)
            nothing
        end,
        precompile=true,
        known_flops=2*64*3
    ))

    # GPU forward benchmarks at different scales
    println(stderr, "GPU Forward")

    for (label, model_fn, hidden) in [("64", tiny_metal_model, 64), ("256", small_metal_model, 256), ("512", medium_metal_model, 512)]
        push!(scenarios, run_scenario(
            id="gpu_forward_$(label)", axis="gpu",
            params=Dict{String,Any}("batch" => 2, "seq_len" => 32, "hidden_dim" => hidden),
            setup_fn=() -> begin
                Random.seed!(SEED); seed_rng!(SEED)
                set_inference_mode!(true)
                model = model_fn()
                input = make_forward_input_mtl(2, 32)
                (model=model, input=input)
            end,
            run_fn=ctx -> begin
                gpu_forward(ctx.model, ctx.input)
                nothing
            end
        ))
    end

    # GPU train benchmarks at different scales
    println(stderr, "GPU Train")

    for (label, model_fn, hidden) in [("64", tiny_metal_model, 64), ("256", small_metal_model, 256), ("512", medium_metal_model, 512)]
        push!(scenarios, run_scenario(
            id="gpu_train_$(label)", axis="gpu",
            params=Dict{String,Any}("batch" => 2, "seq_len" => 8, "hidden_dim" => hidden),
            setup_fn=() -> begin
                Random.seed!(SEED); seed_rng!(SEED)
                set_inference_mode!(true)
                model = model_fn()
                cfg = default_metal_train_config()
                trainer = MetalTrainer(model, cfg)
                input = make_forward_input_mtl(2, 8)
                targets = make_targets_mtl(2, 8)
                (trainer=trainer, input=input, targets=targets)
            end,
            run_fn=ctx -> begin
                gpu_forward_ce_no_readback!(ctx.trainer, ctx.input, ctx.targets)
                nothing
            end
        ))
    end

    # JSON output
    cpu_model = get_cpu_model()
    timestamp = Dates.format(now(Dates.UTC), dateformat"yyyy-mm-ddTHH:MM:SSZ")
    scenarios_json = join(map(scenario_to_json, scenarios), ",\n    ")

    json = """{
  "metadata": {
    "language": "julia",
    "backend": "gpu_metal",
    "language_version": "$(escape_json(string(VERSION)))",
    "os": "$(escape_json(string(Sys.KERNEL)))",
    "cpu_model": "$(escape_json(cpu_model))",
    "timestamp": "$(timestamp)",
    "n_trials": $(N_TRIALS),
    "n_warmup": $(N_WARMUP),
    "seed": $(SEED)
  },
  "scenarios": [
    $(scenarios_json)
  ]
}
"""

    print(json)
    flush(stdout)
    println(stderr, "GPU Benchmark complete. $(length(scenarios)) scenarios.")
    flush(stderr)
end

main()
